### PR TITLE
Data visualizations in the HMIS DQ Tool

### PR DESCRIPTION
**Pull request by eanders**
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

This provides a "By Chart" tab to the HMIS DQ Tool.
The tab covers:
* Client Completeness
* Enrollment Completeness
* Timeliness
* Time in Enrollment

Screenshot 2024-06-07 at 2 01 57 PM

Additionally it provides a PDF download of these charts.

[HMIS Data Quality Tool 2024-06-07 13 24 04.pdf](https://github.com/user-attachments/files/15743337/HMIS.Data.Quality.Tool.2024-06-07.13.24.04.pdf)

## Type of change
[//]: # 'remove options that are not relevant'

- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/hmis_data_quality_tool/app/views/hmis_data_quality_tool/warehouse_reports/reports/_timeliness.haml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+          %p.jLoading Loading
+      .col-sm-6
+        %h3.text-center Time to Exit
+        %h4.text-center Goal: #{pluralize(@report.timeliness_exit_goal, 'day')} or less
+        .time-to-exit-chart.jChart
+          %p.jLoading Loading
+
+= content_for :page_js do
+  :javascript
+    (function($) {
```
Comment (ttoomey) [inline drivers/hmis_data_quality_tool/app/views/hmis_data_quality_tool/warehouse_reports/reports/_timeliness.haml]: Small thing, maybe better to invoke it like this to ensure the dom is ready
```js
$(function() {
      //var time_to_enter = new HmisDqToolTimeToEnter(...
})
```

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/document_exports/report_chart_pdf_export.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        # Ensure we have per-page JS
+        per_page_js: [
+          'hmis_dq_tool_completeness',
+          'hmis_dq_tool_time_to_enter',
+          'hmis_dq_tool_time_in_enrollment',
+        ],
+      }
+    end
+
+    def perform
```
Comment (ttoomey) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/document_exports/report_chart_pdf_export.rb:34]: the perform method seems to be repeated  or similar across 7 different DocumentExports. Perhaps this should be dried up at some point.

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb [hunk 3]
```diff
@@ -380,6 +380,154 @@ def average_days_before_entry
       sum / count
     end
 
+    def average_days_to_enter_entry_date
+      count = enrollments.count
+      sum = enrollments.sum(:days_to_enter_entry_date)
+
+      return 0 if count.zero?
+
+      sum / count
```
Comment (ttoomey) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb]: these are integers. That means the value will be truncated (not rounded). is that expected? (same on line 380, 400, and 480)
```ruby
> (3/2)
=> 1
(3.0/2).round
=> 2
```

I guess I would expect a float here
```suggestion
      sum.to_f / count
```

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb [hunk 4]
```diff
@@ -380,6 +380,154 @@ def average_days_before_entry
...
+
+    def goal_segments
+      @goal_segments ||= goal_config&.active_segments
+    end
+
+    def completeness_target
+      return nil unless goal_segments.present?
+
+      # Return the bottom of the top segment
+      goal_segments.values.map(&:first).max.inspect
```
Comment (ttoomey) [inline drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb]: usually `inspect` is for debugging. Maybe should be `to_s`?

**Review by ttoomey — APPROVED** [review]
Looks good!
* The average calculation in the HmisDataQualityTool::Report seems odd to me but maybe that's the expected behavior, see note
* The export classes and js chart configs follow the established pattern but there's repeated code and maybe opportunities to streamline (perhaps tidy up the js/html glue using stimulus). Could be worth doing if this seems these reports will grow in the future

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb (hunk 4 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb]: Huh, that was just some left-over debugging I guess.  Thank you!

File: drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb (hunk 3 above)
Comment (eanders) [reply drivers/hmis_data_quality_tool/app/models/hmis_data_quality_tool/report.rb]: Good catch, thank you!

**Pull request merged**